### ISSUE TITLE
Efficient encrypted message identification

### ISSUE THREAD

**Issue by s1fr0** [waku-org/nwaku#1081]
# Problem
To reduce metadata leakage and prevent linking encrypted messages exchanged within the same communication, in [35/WAKU2-NOISE](https://rfc.vac.dev/spec/35/) it is suggested that users frequently change `contentTopic`, ideally for every message or at least every 1 round trip communication. Since new content topics are derived deterministically from a secret shared between communicating users, this further allow parties to quickly identify encrypted messages addressed to them and hence drastically reduce the number of required trial-decryptions (as it would happen if messages were exchanged on a public content topic, similarly as happens in [26/WAKU-PAYLOAD v1](https://rfc.vac.dev/spec/26/#decoding-a-message)). 

In order to retrieve messages addressed to a specific content topic, users can either:
- sieve all messages exchanged on the parent pubsub topic and locally filter the ones having the content topic of their interest;
- make a filter query to store nodes, but the latter might flag the user making the query as one of the users involved in the corresponding encrypted communication, thus partially invalidating the anonymity goal the frequent content topic change was trying to achieve.

This issue partially addresses https://github.com/vacp2p/research/issues/131, by proposing an alternative mechanism to frequent content topic change.

# Possible Solution/tradeoff
A possible tradeoff could be given by adding a `message-id` field to the Waku message [payload v2](https://rfc.vac.dev/spec/35/#abnf), so that
- handshake and after-handshake messages are exchanged under an assigned `contentTopic`, e.g. `/{application-name}/{application-version}/wakunoise/1/sessions-{shard-id}/proto`;
- users can make filter queries only for such content topic, thus keeping an higher level of anonymity (only `shard-id` leaked);
- users will perform decryption only on messages having (one of) the expected `message-id`;
- `message-id` is derived deterministically from a secret shared between communicating users, and can in principle be changed at every message;
- sender and recipient can keep updated a buffer of `message-id` to sieve (i.e., the next 50 not yet seen) and can thus further identify if one or more messages were lost;

We note that the the default cipher for payload v2 messages is `ChaChaPoly`, which supports `additional data`: hence the ciphertext can be authenticated by passing the `message-id` as additional data, so that an attacker cannot change the (non-encrypted) `message-id` in a Waku message without invalidating the decryption of the ciphertext.

# Acceptance criteria
We look for a solution which is efficient in terms of trial-decryption, preserves as much as possible user anonymity while being practical given the gossip/sub protocol.

@jm-clius @LNSD @staheri14 @oskarth for a feedback if the proposed solution can make sense in Waku2, or if you see an alternative way

**Comment by jm-clius** [comment]
I think this solution would work, but I'd like to suggest something else for a number of reasons :)
- we have good reason to try to keep the `WakuMessage` as simple as possible. For one, we want to move the WAKU2-MESSAGE spec to `stable` soon, as it's a prerequisite for almost all other specs to be `stable`. As far as possible we should avoid adding fields that are only of interest to a subset of users (these usually go into the specific application).
- information that has as goal informing the Waku node that it might be interested in the content of the message (e.g. "you should decrypt this") falls more in the domain of "content filtering" than extra message information.

For this reason I'd propose rather overloading the `contentTopic` field more and adapting the `filter` protocol to allow for partial matching (which we should do in any case).
- e.g. messages are exchanged on `/{application-name}/{application-version}/wakunoise/1/sessions-{shard-id}/{message-id}/proto`
- nodes subscribe to everything under `/{application-name}/{application-version}/wakunoise/1/sessions-{shard-id}/*`.
- nodes only decrypt messages based on the {message-id} component in the content topics
- there's no cost to adding more content topics, as long as the filters can be coarse enough to avoid having to constantly update subscriptions

WDYT?

One more note: we should come up with a term other than `message-id` here, as the concept [already exists on the libp2p layer ](https://github.com/libp2p/specs/tree/098dab3403a3c608ddd77d89a3a04b2bc23be7c8/pubsub#message-identification).

**Comment by jm-clius** [comment]
> Just to clarify: I would add the `message-id` field to the [Noise payload v2](https://rfc.vac.dev/spec/35/#abnf), not to the Waku message. In the latter, indeed, the extra field would be transparent since will be serialized into its `payload` field.

Oh, you're right! I didn't read properly. :D

> Given this detail, the two solutions seems to be equivalent in practice since all messages under `/{application-name}/{application-version}/wakunoise/1/sessions-{shard-id}/*` has to be downloaded anyway. But maybe we can make more efficient local filtering by checking the shorter WakuMessage `contentTopic` field from the protobuffer rather than the longer `payload` field (especially if padded by default), so I would prefer yours.

Indeed. This feels like a subset of the content topic and may be more efficient this way, but indeed, the two solutions are equivalent enough that I won't have strong objections to either.

**Comment by staheri14** [comment]
@s1fr0 
A few questions to better understand the problem
- After a handshake is done let's say between Alice and Bob, Alice and Bob (by default) will use a single content topic X (derived from their handshake) to exchange their encrypted messages. However, using one single content topic X will be against unlinkability of chat messages. Thus, we want a solution in which the content topic of a 1:1 communication does not allow such linkability, right?
- There is also another issue with regard to retrieving historical encrypted messages, where the querier e.g., Alice needs to reveal its desired content topic X (or content topics x_1, x_2, etc. in case there are multiple content topics per 1:1 communication) to the store node. So, we are looking for a solution to lower this leakage, right?

**Comment by oskarth** [comment]
Probably makes sense to get feedback from Status CCs here as they have experience with this, people like @cammellos @iurimatias @richard-ramos - even if this ultimately isn't useful to Status app right now their expertise is is valuable in terms of evaluating a solution for this.

**Comment by oskarth** [comment]
Which milestone is this part of?

**Comment by fryorcraken** [comment]
As mentioned in https://github.com/vacp2p/research/pull/134#issuecomment-1232464966

To access messages on a given content topic, a number of action may be required (do a store query for given content topic, subscribe to content topic).

As I understand, a message id is generated from information last received by the other party. It then may create race conditions and/or inefficiency if, let's say, a filter subscribe needs to be done before the next noise session round.

Hence, for the final solution I would recommend to stick to one content topic, maybe something quite broad to avoid metadata leakage (and we already have the sharding option in place if the number of messages to process on the given content topic is to high), and then having a field on the message, such as id or something else, that enables a party to not have to decrypt all messages in the content topic.
